xapian-maintainer-tools/profiling/profiler.c

```diff
@@ -0,0 +1,189 @@
+/** @file profiler.c
+ * @brief C file implementing IO logging
+ */
+/* Compile the profiler with:
+ * gcc -shared -fPIC -o profiler.so profiler.c -ldl
+ *
+ * To use it with a process, use:
+ * LD_PRELOAD=/path/to/profiler.so /path/to/executable
+ *
+ * Running this library will produce an strace.log file,
+ * which can be fed to strace-analyse to produce a log, by running:
+ * ./strace-analyse strace.log
+ */
+
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+
+#define MAX_CALL_LEN 5000
+
+// function for logging calls
+
+int reset = 0;
+
+void logcall(const char *call)
+{
+    FILE *file_ptr;
+    if (!reset) {
+	file_ptr = fopen("./strace.log", "w");
+	reset = 1;
+    } else {
+	file_ptr = fopen("./strace.log", "a");
+    }
+    fprintf(file_ptr, "%s", call);
+    fclose(file_ptr);
+}
+
+// wrapper for open()
+
+typedef int(*real_open_t)(const char *, int, ...);
+
+int open(const char *pathname, int flags, ...)
+{
+    int fd;
+    if (flags & O_CREAT || flags & O_TMPFILE) {
+	va_list args_ptr;
+	va_start(args_ptr, flags);
+	fd =
+	    ((real_open_t)dlsym(RTLD_NEXT, "open"))
+	    (pathname, flags, va_arg(args_ptr, mode_t));
```

Comment:
From `man va_start`:

```
   va_end()
       Each  invocation  of va_start() must be matched by a corresponding invocation of va_end() in the same function.  
```

You need to call `va_end()` here...

Reply:
I apologize. Updated accordingly.